File: SpeedGrader/SpeedGrader/Router.m
//
// Copyright (C) 2016-present Instructure, Inc.
//   
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, version 3 of the License.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.
//

#import <objc/runtime.h>

#import "Router.h"
#import "Router+Routes.h"

@interface Router ()
@property NSMutableDictionary *routes;
@property NSNumberFormatter *numberFormatter;
@end

@implementation UIViewController (Routing)

- (void)applyRoutingParameters:(NSDictionary *)params {
    [params enumerateKeysAndObjectsUsingBlock:^(NSString *key, id obj, BOOL *stop) {
        NSString *selectorString;
        
        objc_property_t prop = class_getProperty([self class], [key UTF8String]);
        char *setterName = property_copyAttributeValue(prop, "S");
        if (setterName) {
            selectorString = [NSString stringWithFormat:@"%s", setterName];
        }
        else {
            const char *selName = [key UTF8String];
            selectorString = [NSString stringWithFormat:@"set%c%s:", toupper(selName[0]), selName+1];
        }
        
        SEL selector = NSSelectorFromString(selectorString);
        if ([self respondsToSelector:selector]) {
            [self setValue:obj forKey:key];
        }
    }];
}

@end

@implementation Router

- (id)init {
    self = [super init];
    if (self) {
        _routes = [NSMutableDictionary new];
        _numberFormatter = [NSNumberFormatter new];
        [self configureInitialRoutes];
    }
    return self;
}

+ (Router *)sharedRouter {
    static Router *_sharedRouter = nil;
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        _sharedRouter = [Router new];
    });
    
    return _sharedRouter;
}

#pragma marks - Defining Routes

- (void)addRoute:(NSString *)route handler:(RouteHandler)handler {
    self.routes[route] = handler;
}

- (void)addRoute:(NSString *)route forControllerClass:(Class)controllerClass {
    self.routes[route] = controllerClass;
}

- (void)addRoutesWithDictionary:(NSDictionary *)routes {
    [routes enumerateKeysAndObjectsUsingBlock:^(id key, id obj, BOOL *stop) {
        // if it's a class, make sure it's a subclass of UIViewController
        if (class_isMetaClass(object_getClass(obj))) {
            if (![obj isSubclassOfClass:[UIViewController class]]) {
                [NSException raise:@"Not a subclass of UIViewController" format:@"Must pass in either a subclass of UIViewController or handler block"];
            }
        }
        
        self.routes[key] = obj;
    }];
}

#pragma marks - Dispatching

- (UIViewController *)controllerForClass:(Class)cls params:(NSDictionary *)params {
    UIViewController *viewController = [cls new];
    [viewController applyRoutingParameters:params];
    return viewController;
}

#pragma marks - Primary iPad Routing Methods
- (UIViewController *)controllerForHandlingBlockFromURL:(NSURL *)url {
    __block UIViewController *returnedController;
    
    [self matchURL:url matchHandler:^(NSDictionary *params, id classOrBlock) {
        if (class_isMetaClass(object_getClass(classOrBlock))) { // it's a class
            returnedController = [self controllerForClass:classOrBlock params:params];
        } else {
            UIViewController *(^blockForPath)(NSDictionary *, id) = classOrBlock;
            returnedController = blockForPath(params, nil);
        }
    }];
    
    return returnedController;
}

- (UIViewController *)routeFromController:(UIViewController *)sourceController toURL:(NSURL *)url {

    UIViewController *destinationViewController = [self controllerForHandlingBlockFromURL:url];
    
    if (!destinationViewController && self.fallbackHandler) {
        self.fallbackHandler(url, sourceController);
        return nil;
    }
    
    if ([sourceController isKindOfClass:[UINavigationController class]]) {
        [((UINavigationController *)sourceController) pushViewController:destinationViewController animated:YES];
    } else {
        [sourceController.navigationController pushViewController:destinationViewController animated:YES];
    }
    
    return destinationViewController;
}

# pragma marks - Parsing

- (BOOL)matchURL:(NSURL *)url matchHandler:(void(^)(NSDictionary *params, id classOrBlock))matchHandler {
    NSArray *urlComponents = url.pathComponents;
    
    // strip /api/v1/ if it's there
    if (urlComponents.count > 2 && [urlComponents[1] isEqualToString:@"api"] && [urlComponents[2] isEqualToString:@"v1"]) {
        urlComponents = [urlComponents subarrayWithRange:NSMakeRange(3, urlComponents.count - 3)];
    }
    else if (urlComponents.count > 0) { // strip slash
        urlComponents = [urlComponents subarrayWithRange:NSMakeRange(1, urlComponents.count -1)];
    }

    __block NSMutableDictionary *params;
    __block BOOL matchFound = NO;
    
    [self.routes.allKeys enumerateObjectsUsingBlock:^(NSString *key, NSUInteger idx, BOOL *stop) {
        NSURL *routeURL = [NSURL URLWithString:key];
        NSArray *routeComponents = [routeURL.pathComponents subarrayWithRange:NSMakeRange(1, routeURL.pathComponents.count -1)];
        params = [NSMutableDictionary new];
        
        if (urlComponents.count != routeComponents.count) {
            return; // can't possibly match
        }
        
        for (NSUInteger i = 0; i < routeComponents.count; i++) {
            if ([routeComponents[i] hasPrefix:@":"]) { // save params in case this is a match
                NSString *paramKey = [routeComponents[i] substringFromIndex:1];
                NSString *param = urlComponents[i];
                NSNumber *number = [self.numberFormatter numberFromString:param];
                if (number) {
                    params[paramKey] = number;
                }
                else {
                    params[paramKey] = param;
                }
            }
            else if (![urlComponents[i] isEqualToString:routeComponents[i]]) {
                return; // not a match, continue to next key
            }
        }
        
        matchFound = YES;
        if (self.canvasAPI) {
            params[@"canvasAPI"] = self.canvasAPI;
        }
        matchHandler(params, self.routes[key]);
        *stop = matchFound;
    }];
    
    if(!matchFound){
        NSLog(@"no registered route for URL %@", url);
    }
    
    return matchFound;
}

@end

File: Parent/Parent/Router/Router.swift
//
// Copyright (C) 2016-present Instructure, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, version 3 of the License.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.
//

import UIKit
import CanvasCore

typealias RouteHandler = (_ params: [String : Any]?) -> UIViewController
typealias FallbackHandler = (_ url: URL?) -> UIViewController?

class Router {
    
    var session: Session?
    var window: UIWindow?
    
    static let sharedInstance = Router()
    
    fileprivate var fallbackHandler : FallbackHandler = { url in    // Private for now.  No need to open this up until it's needed
        guard let url = url else {
            return UIViewController()
        }
        
        if url.scheme == "parent-courses" {
            return nil
        }
        
        // TODO: Setup a SafariViewController here
        return UIViewController()
    }
    
    fileprivate var routes = [String : RouteHandler]()
    fileprivate var numberFormatter = NumberFormatter()
    
    // ---------------------------------------------
    // MARK: - Adding Routes
    // ---------------------------------------------
    /**
     Adds a Route Template handler to our route map
    
     Route Templates follow a URL path scheme with variables inserted with a ":" prefix
    
     For example: If a route is added with the following route template - "user/assignments/:assignmentID" then
     a matching URL for this parameter would be something like "scheme://domain.com/user/assignments/234"
    
     - Parameters
        - route: URL Path Template for this handler
        - handler: completion block returning a view controller created from parameters in the URL Path Template
    */
    func addRoute(_ route: String, handler: @escaping RouteHandler) {
        routes[route] = handler
    }
    
    /**
     Adds multple routes to the route map
     
     Route templates follow the same rules as routes from the 'addRoute' method
     
     - Parameters
        - newRoutes: mapping similar to our map that key a Route Handler to a Route Template
    */
    func addRoutesWithDictionary(_ newRoutes: [String : RouteHandler]) {
        for route in newRoutes.keys {
            routes[route] = newRoutes[route]
        }
    }
    
    // ---------------------------------------------
    // MARK: - Dispatching
    // ---------------------------------------------
    /**
     Fetches the RouteHandler for a given URL if available, and then applies the given parameters to return a resulting  UIViewController
     
     - Parameter url: The URL of the ViewController you are trying to route to
    
     - Returns: UIViewController if we have a mapping that matches the URL provided.  Returns nil if no mapping is available.
    */
    func viewControllerForURL(_ url: URL?) -> UIViewController? {
        guard let url = url else { return nil }
        
        return matchURL(url)
    }
    
    
    /**
     Routes from a specific view controller to the URL provided.
     
     - Parameters
        - fromController: View Controller initializing the
        - toURL: the URL you want to route to.
        - animated: Should the transition be animated?
        - modal: Is the viewController modal?
     
     - Returns: The view controller that was routed to
     */
    @discardableResult
    func route(_ fromController: UIViewController, toURL: URL, animated: Bool = true, modal: Bool = false) -> UIViewController {
        guard let viewController = matchURL(toURL) else {
            // Fallback view controller here
            // Return Fallback View Controller
            return UIViewController()
        }
        
        fromController.transitionToViewController(viewController, animated: animated, modal: modal)
        return viewController
    }
    
    /**
     Routes from a specific view controller to the URL provided.
     
     - Parameters
     - window: Window containing viewControllers
     - toRootViewController: ViewController to route to
     - animated: Should the transition be animated?
     */
    func route(_ window: UIWindow, toRootViewController viewController: UIViewController, animated: Bool = true) {
        let navController = UINavigationController(rootViewController: viewController)
        navController.isNavigationBarHidden = true
        let animationDuration = animated ? 0.5 : 0.0
        UIView.transition(with: window, duration: animationDuration, options: UIViewAnimationOptions(), animations: { _ in
            window.rootViewController = navController
            }, completion: nil)
    }
    
    
    // ---------------------------------------------
    // MARK: - URL Parsing
    // ---------------------------------------------
    /**
     Matches a URL to the existing Route Templates in our route mapping.  It then executes the handler with the parameters provided.  Additionally it adds a session if one is available in the "session" variable of the parameters to be handled by the view controller.
    
     Parameter url: url that hopefully matches a Route Template in the mapping.
    
     Returns: A View Controller created from the handler if possible, if not it returns nil
    */
    fileprivate func matchURL(_ matchURL: URL) -> UIViewController? {
        
        // Without a path, how can we route anywhere
        var urlComponents: [String] = []
        let realPath = matchURL.path.replacingOccurrences(of: "api/v1/", with: "")
        urlComponents = URL(fileURLWithPath: realPath).pathComponents.filter { $0 != "/" }

        guard urlComponents.count > 0 else { return nil }
        
        var matchingRoute : String? = nil
        var params = [String: Any]()
        for route in routes.keys {
            // Verify that an invalid route wasn't passed in
            guard let routePathComponents = URL(string: route)?.pathComponents, routePathComponents.count > 0 else {
                continue
            }
            
            // Can't match if the size is different
            guard routePathComponents.count == urlComponents.count else {
                continue
            }
            
            var componentsMatch = true
            for (index, component) in routePathComponents.enumerated() {
                if component.hasPrefix(":") {
                    // remove the ":" from the path
                    let parameterKey = component.substring(from: component.index(component.startIndex, offsetBy: 1))
                    let parameter = urlComponents[index]
                    if let number = self.numberFormatter.number(from: parameter) {
                        params[parameterKey] = number
                    } else {
                        params[parameterKey] = parameter
                    }
                } else if (urlComponents[index] != routePathComponents[index]) {
                    componentsMatch = false
                    break
                }
            }
            
            guard componentsMatch == true else {
                continue
            }
            
            matchingRoute = route

            params["route"] = route
            params["url"] = matchURL
            if let session = session {
                params["session"] = session
            }
            for queryItem in matchURL.allQueryItems {
                params[queryItem.name] = queryItem.value
            }
        }
        
        if let route = matchingRoute, let handler = routes[route] {
            return handler(params)
        }
        
        return nil
    }
    
}

// ---------------------------------------------
// MARK: - UIViewController Transitioning Delegates
// ---------------------------------------------
extension UIViewController {
    /**
     Currently this application only supports UINavigationController transitions.  Breaking this out in this manner though allows us to have other transition controllers in the future if we need to.
     
     - Parameters
        - viewController: the view controller you wish to present to the user
        - animated: should we animate the transition?
        - modal: do you want the view to be specifically presented as a modal
     
    */
    func transitionToViewController(_ viewController: UIViewController, animated: Bool, modal: Bool) {
        // for now we're just going to route using a navigationController or Modal
        if modal {
            present(viewController, animated: animated, completion: { })
        } else {
            guard let navigationController = navigationController else {
                present(viewController, animated: animated, completion: { })
                return
            }

            navigationController.pushViewController(viewController, animated: animated)
        }
    }
}
